### ISSUE TITLE
Types that point to each other generate C code that won't compile

### ISSUE THREAD

**Issue by nsanmartin** [austral/austral#371]
This issue is similar to #358 fixed by #360. The compiler accepts the following input

```
import Austral.Memory (
    Pointer
);

module body Test is
    pragma Unsafe_Module;

    record Bar: Free is
        fooptr: Pointer[Foo];
    end;

    record Foo: Free is
        barptr: Pointer[Bar];
    end;

    function main(): ExitCode is
        return ExitSuccess();
    end;
end module body.
```

but the generated code won't compile:
```

Standard error:
    /tmp/austral_fff442.c:3372:17: error: unknown type name 'mono_13'
         3372 | typedef struct {mono_13* barptr;} mono_12;
              |                 ^~~~~~~


```

The fix for this may be to generate forward declarations for all types before theirs definitions as @eudoxia0 [suggested](https://github.com/austral/austral/pull/360#issuecomment-1374861617)

**Comment by eudoxia0** [comment]
Huh, I thought I was already doing this here: https://github.com/austral/austral/blob/master/lib/CodeGen.ml#L593-L609

**Comment by eudoxia0** [comment]
Let me at least add a test for this.

**Comment by eudoxia0** [comment]
This is fixed now: #372 

Thanks for reporting!
